torch/_inductor/fx_passes/quantization.py

```diff
@@ -2539,6 +2598,175 @@ def _register_qlinear_weight_prepack():
         )
 
 
+def _register_linear_dynamic_fp16_weight_prepack_pass(
+    pattern,
+    pass_number,
+    input_dim_exceeds_two=False,
+    input_contiguous=True,
+):
+    def _extra_check_fn(match: Match):
+        return match.kwargs["dtype_fp16"] == torch.float16
+
+    @register_freezing_graph_pattern(
+        pattern,
+        extra_check=_extra_check_fn,
+        pass_number=pass_number,
+    )
+    def linear_dynamic_fp16_weight_prepack(match: Match, *args, **kwargs):
+        """
+        Match the pattern:
```

Comment:
this is complicated, is there any way we can get the pattern from tracing a higher level pattern?

Reply:
Thanks for the suggestion. Do you have any examples as references? Thanks.

Reply:
for example this is how we get pattern for qat: https://github.com/pytorch/pytorch/blob/7dfb439a2af2c9f084fed2fcd046b7bd1ec54dbf/torch/ao/quantization/pt2e/qat_utils.py#L676

Reply:
Thanks. However, here we use the pattern matcher in Inductor and the pattern is described by things like `CallFunction`. So, looks like we cannot generate the pattern from tracing. And now we use hand-written patterns as defined in `_generate_linear_dynamic_fp16_pattern` above. So, looks like we cannot generate the pattern from tracing.

Reply:
Hi @jerryzh168 Do you have more comments on this? Thanks.

Reply:
OK, sounds good, I feel if inductor can output the intermediate graph you might still be able to define the higher level pattern and trace it to get the inductor level pattern though, I'm not sure if there is any blockers for this though

Reply:
but looks good to land now

Reply:
Thanks. I agree that it may be simpler if the pattern can be generated via tracing. Does anyone know if there are such functionalities in Inductor?

Reply:
[SDPA patterns in Inductor](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/fuse_attention.py) are defined with PyTorch API, and then their Inductor IR patterns can be serialized. Are you referring to something like this?

Reply:
@sanchitintel Thanks for the pointer. Is it done by tracing or handwriting?

Reply:
Individual patterns defined in `fuse_attention.py` with torch API are traced to produce serialized patterns of the kind you were alluding to (a `CallFunction` that may be nested). This approach prevents writing nested `CallFunction`s manually.

Reply:
How to do the tracing? Thanks.

Reply:
Please refer to this code-flow as an example -  https://github.com/pytorch/pytorch/blob/b31d3b2f418229c9267594b49ffc4e276b76e580/torch/_inductor/fx_passes/fuse_attention.py#L912-L914

You could try playing around with this code, as it may help narrow down on what underlying API you could use for your use-case to generate nested `CallFunction`s corresponding to high-level patterns written with torch API.

Thanks!

Reply:
It looks quite complicated. I will try to understand it later. Thanks.